lib/srv/desktop/rdp/rdpclient/src/network_client.rs

```diff
@@ -0,0 +1,149 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * /
+ */
+use std::future::Future;
+use std::net::{IpAddr, Ipv4Addr};
+use std::pin::Pin;
+
+use ironrdp_connector::sspi::generator::NetworkRequest;
+use ironrdp_connector::sspi::network_client::NetworkProtocol;
+use ironrdp_connector::{custom_err, ConnectorResult};
+use ironrdp_tokio::AsyncNetworkClient;
+use reqwest::Client;
+use sspi::{Error, ErrorKind};
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+use tokio::net::{TcpStream, UdpSocket};
+use url::Url;
+
+pub(crate) struct ReqwestNetworkClient {
+    client: Option<Client>,
+}
+
+impl AsyncNetworkClient for ReqwestNetworkClient {
+    fn send<'a>(
+        &'a mut self,
+        request: &'a NetworkRequest,
+    ) -> Pin<Box<dyn Future<Output = ConnectorResult<Vec<u8>>> + 'a>> {
+        Box::pin(async move {
+            match &request.protocol {
+                NetworkProtocol::Tcp => self.send_tcp(&request.url, &request.data).await,
+                NetworkProtocol::Udp => self.send_udp(&request.url, &request.data).await,
+                NetworkProtocol::Http | NetworkProtocol::Https => {
+                    self.send_http(&request.url, &request.data).await
+                }
+            }
+        })
+    }
+}
+
+impl ReqwestNetworkClient {
+    pub(crate) fn new() -> Self {
+        Self { client: None }
+    }
+}
+
+impl ReqwestNetworkClient {
+    async fn send_tcp(&self, url: &Url, data: &[u8]) -> ConnectorResult<Vec<u8>> {
+        let addr = format!(
+            "{}:{}",
+            url.host_str().unwrap_or_default(),
+            url.port().unwrap_or(88)
+        );
+
+        let mut stream = TcpStream::connect(addr)
+            .await
+            .map_err(|e| Error::new(ErrorKind::NoAuthenticatingAuthority, format!("{:?}", e)))
+            .map_err(|e| custom_err!("failed to send KDC request over TCP", e))?;
+
+        stream
+            .write(data)
+            .await
+            .map_err(|e| Error::new(ErrorKind::NoAuthenticatingAuthority, format!("{:?}", e)))
+            .map_err(|e| custom_err!("failed to send KDC request over TCP", e))?;
+
+        let len = stream
+            .read_u32()
+            .await
+            .map_err(|e| Error::new(ErrorKind::NoAuthenticatingAuthority, format!("{:?}", e)))
+            .map_err(|e| custom_err!("failed to send KDC request over TCP", e))?;
+
+        let mut buf = vec![0; len as usize + 4];
+        buf[0..4].copy_from_slice(&(len.to_be_bytes()));
+
+        stream
+            .read_exact(&mut buf[4..])
+            .await
+            .map_err(|e| Error::new(ErrorKind::NoAuthenticatingAuthority, format!("{:?}", e)))
+            .map_err(|e| custom_err!("failed to send KDC request over TCP", e))?;
```

Comment:
Try not to use the same error message in multiple places, it makes troubleshooting difficult.

Reply:
I've added separate error messages